### DISCUSSION TITLE
Discussion: Permit static lambda assignment to function pointers

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#6746] [Language Ideas]
It should be safe to pass a static lambda to a function pointer.

```cs
public static unsafe class C {
    static bool Any<T,A>(this T[] array, delegate*<T, A, bool> predicate, A arg) => false;

    public static void M(int a, int b) {  
        new []{1}.Any(&static (item, t) => item == t.a, (a, b));
    }

    // instead of
    public static void M(int a, int b) {  
        static bool predicate(int item, (int a, int b) t) => item == t.a;   
        new []{1}.Any(&predicate, (a, b));
    }
}
```
Otherwise you must fallback to using local function and declare everything explicitly.

**Comment by RikkiGibson** [comment]
This requires a new feature which amends these parts of the ['static lambdas' speclet](https://github.com/dotnet/csharplang/blob/master/proposals/static-lambdas.md):

> No guarantee is made as to whether a `static` lambda definition is emitted as a `static` method in metadata. This is left up to the compiler implementation to optimize.

We will now need to guarantee that a 'static' lambda is emitted as 'static' if its address is taken.

> Removing the `static` modifier from a lambda in a valid program does not change the meaning of the program.

Removing the 'static' modifier in a context where it is converted to function pointer will introduce a compile-time error.

**Comment by 333fred** [comment]
> We will now need to guarantee that a 'static' lambda is emitted as 'static' if its address is taken.

Presuming, of course, that we would want to make those changes. We explicitly did not make any guarantees about how `static` lambdas are emitted, and I don't find this feature compelling enough to change that.

**Comment by alrz** [comment]
> I don't find this feature compelling enough to change that.

To me, this feels like a feature gap in the language, something that you would expect to be possible but it doesn't work merely because it's not implemented. But I'd agree that it's not very common because function pointers hasn't been shipped yet ;)

**Comment by 333fred** [comment]
> something that you would expect to be possible but it doesn't work merely because it's not implemented

That's not how I think about this at all. Function pointers aren't a common use case. They require unsafe, and getting a pointer requires taking the address of a static method. I wouldn't expect them to work with anything else.

**Comment by alrz** [comment]
>They require unsafe, and getting a pointer requires taking the address of a static method.

Within that context, one could decide to use a function pointer instead of a delegate for performance reasons. Then they need to use a local function anywhere they previously used a (static) lambda. Now we're back where we felt the need for static lambdas in the first place (while static local functions were available).

**Comment by alrz** [comment]
I understand that technically a proper method is required to acquire an address, but this would be a convenient syntactic sugar much like what has been done for `in` parameters -- create a local on-the-fly and pass a ref of that if the current argument is not addressable.

**Comment by AraHaan** [comment]
Tbh why not just make the static lambda a private static member in the type then convert that to a function pointer and return it where you need to use it instead of using a ``local function``?

**Comment by RikkiGibson** [comment]
Wanted to address https://github.com/dotnet/csharplang/discussions/6111#discussioncomment-2723273 in the original issue (cc @HaloFour)

> I think that would also tie the compiler to a specific implementation of how static lambdas are emitted. They currently are instance methods on a singleton instance, but that would be incompatible with the function pointer.

```cs
Func<int, int> del = static int (int x) => x;
delegate*<int, int> funcPtr = &static int (int x) => x;
```

A lambda is only used once, so it seems reasonable that the way it is used could affect the way it is lowered. Why couldn't the compiler emit a non-`static` method for `del`, and a `static` method for `funcPtr`, for example?

**Comment by HaloFour** [comment]
@RikkiGibson 

> A lambda is only used once, so it seems reasonable that the way it is used could affect the way it is lowered. 

I'd like to see more of that in general, like if a lambda is only invoked from within the method in which it is declared the compiler could emit it as a local function instead, save an allocation or two.

**Comment by alrz** [comment]
>the compiler could emit it as a local function instead, save an allocation or two.

I'm probably missing something but how it could possibly save any allocations?

**Comment by RikkiGibson** [comment]
I think @HaloFour's suggestion was to adjust lowering/emit in some cases so that no local variable is introduced and no delegate is actually allocated for the lambda.

```cs
Func<int, int> multiply = x => x * 2;
Console.Write(multiply(2));


// equivalent to
int multiply(int x) => x * 2;
Console.Write(multiply(2));
```

**Comment by alrz** [comment]
Got it. I thought it has something to do with `&static` lambdas. The only advantage of that I think is type inference.

**Comment by timcassell** [comment]
Considering this spec

> Removing the static modifier from a lambda in a valid program does not change the meaning of the program.`

I would find it useful to be able to directly assign a lambda to a function pointer without `&` or `static`, as long as no variables are captured.

My use case is, I have a library that supports old versions of Unity (C# 4), and I conditionally compile with newer C# language features for better performance.

A simplified example similar to what I'm doing:

```cs
internal delegate void ConvertDelegate<T>(object asyncResult, ref T storage, int index);

internal Merger<T> : IAsyncResult<T>, IIndexedCallback
{
    internal T _result;
    private ConvertDelegate<T> _converter;
    
    internal Merger(ConvertDelegate<T> converter)
    {
        _converter = converter;
    }
    
    void IIndexedCallback.Callback(object asyncResult, int index)
    {
        _result = _converter(asyncResult, ref _result, index);
    }
    
    T IAsyncResult<T>.Result { get { return _result; } }
}

public static AsyncResult<ValueTuple<T1, T2>> Merge<T1, T2>(AsyncResult<T1> result1, AsyncResult<T2> result2)
{
    var merger = new Merger<ValueTuple<T1, T2>>((object asyncResult, ref ValueTuple<T1, T2> storage, int index) =>
    {
        if (index == 0)
        {
            storage.Item1 = ((IAsyncResult<T1>) asyncResult).Result;
        }
        else
        {
            storage.Item2 = ((IAsyncResult<T2>) asyncResult).Result;
        }
    });
    result1.AddIndexedListener(merger, 0);
    result2.AddIndexedListener(merger, 1);
    return new AsyncResult<ValueTuple<T1, T2>>(merger);
}

...

public static AsyncResult<ValueTuple<T1, T2, T3, T4>> Merge<T1, T2, T3, T4>(AsyncResult<T1> result1, AsyncResult<T2> result2, AsyncResult<T3> result3, AsyncResult<T4> result4)
{
    ...
}
```

And it would be very convenient for me if I could just change the `Merger` implementation to this so that the lambda expressions wouldn't have to change to fit function pointers and reduce duplicate code:

```cs
#if CSHARP_12_OR_GREATER // Or whichever language version
    private delegate*<object, ref T, int, void> _converter;
    
    internal Merger(delegate*<object, ref T, int, void> converter)
    {
        _converter = converter;
    }
#else
    private ConvertDelegate<T> _converter;
    
    internal Merger(ConvertDelegate<T> converter)
    {
        _converter = converter;
    }
#endif
```

I know my use case is super niche and this may be unlikely to happen, but I figured I should give my input here.

**Comment by MichalPetryka** [comment]
Being able to pass static lambdas to function pointers would be nice, but I must warn you: older Unity versions don't always handle `calli` properly, especially with IL2CPP.

**Comment by timcassell** [comment]
> Being able to pass static lambdas to function pointers would be nice, but I must warn you, older Unity versions don't always handle `calli` properly, especially with IL2CPP.

That's why I use conditional compilation for older Unity versions to just use the regular delegate.

**Comment by MichalPetryka** [comment]
```C#
delegate* unmanaged[Cdecl]<int, float> d = &static (i) => (float)i;
```
This could also implicitly use UnmanagedCallersOnly when encountering an assignment to an unmanaged pointer, would be pretty useful with interop scenarios now that I think about it.

**Comment by MichalPetryka** [comment]
This will become significantly more useful for pointers to managed methods with me fixing https://github.com/dotnet/runtime/issues/44610 (currently still working on it), since then function pointers become a mostly overhead-less way to pass things to call to other methods.

**Reply by jaredpar** [reply]
I understand that will make function pointers a more efficient way of passing a _call back_ but I don't agree that makes this proposal more or less useful. Basically removing that overhead doesn't change the role of function pointers in the ecosystem, it just makes them a bit faster. While that may cause a shift in how many are used, that shift doesn't seem likely to be so vast that we should consider adding more ways to create them into the language. If it does though, then we can add more emphasis to this proposal.

**Reply by jkotas** [reply]
It would be more useful to spend time on making the existing static lambdas more efficient, so that people do not need to resort to function pointers for optimizations. The problem is that the IL for static lambdas is complicated and it is hard for the JIT to reverse-engineer the indent reliably. Can we introduce new BCL APIs that would be used by Roslyn when available and that would be treated as intrinsic by the JIT to fix this problem? What would it take for the delegates that are backing static lambdas to be created as frozen delegates? In the limit, the difference between function pointer callback and equivalent static lambda callback should be only about two processor instruction (when the callback is not inlined) and the nothing (when the callback is inlined).

**Reply by jaredpar** [reply]
Definitely happy to emit lambdas in a better way to help JIT identify them better.

**Reply by CyrusNajmabadi** [reply]
> Can we introduce new BCL APIs that would be used by Roslyn when available and that would be treated as intrinsic by the JIT to fix this problem?

This seems like an interesting and useful pattern for future optimizations.  Give the compilers the entrypoint to do the work.  And then the jit/runtime can always ensure that things it can always detect things since it owns providing that api and consuming it in lockstep.

**Reply by MichalPetryka** [reply]
AFAIR, no BCL API would be needed for delegates, Roslyn would need to cache them in `static readonly` fields instead, example:
```C#
    private sealed class <>c
    {
        public static readonly <>c<>9 = new <>c();

        public static Action <>9__0_0;

        internal void <action>b__0_0()
        {
        }
    }

    [System.Runtime.CompilerServices.NullableContext(1)]
    public static Action action()
    {
        return <>c.<>9__0_0 ?? (<>c.<>9__0_0 = new Action(<>c.<>9.<action>b__0_0));
    }
```
would need to be emitted as:
```C#
    private sealed class <>c
    {
        public static readonly <>c<>9 = new <>c();

        public static readonly Action <>9__0_0 = new Action(<>c.<>9.<action>b__0_0);

        internal void <action>b__0_0()
        {
        }
    }

    [System.Runtime.CompilerServices.NullableContext(1)]
    public static Action action()
    {
        return <>c.<>9__0_0;
    }
```

The only potential issues would be when a user has thousands of lambdas within a single type, we'd then need to split the generated type up in chunks.

**Reply by jaredpar** [reply]
>would need to be emitted as ...

This does mean that we end up allocating more `Delegate` instances during program startup. 

> The only potential issues would be when a user has thousands of lambdas within a single type, we'd then need to split the generated type up in chunks.

These closures are generated per scope, not per containing type. So hard to see where we get to the thousands case except in pathological code.

**Reply by alrz** [reply]
> This does mean that we end up allocating more `Delegate` instances during program startup.
> 

Is there any concern with emitting a new static class per each cached delegate?

**Reply by timcassell** [reply]
> This does mean that we end up allocating more `Delegate` instances during program startup.

AFAIK statics are initialized the first time the class is accessed, not at program startup. Is that not true with NativeAOT?

Also, I'm curious why the generated class contains a class instance, and doesn't just create the delegate from a static function. Is there a performance difference with using an instance method vs a static function? And would that make a difference to the JIT?

**Reply by jaredpar** [reply]
> AFAIK statics are initialized the first time the class is accessed, not at program startup. 

Correct but developers will almost certainly leave us feedback on the fact that we're now allocating delegates eagerly. Customers are very sensitive to how lambdas are emitted and will certainly send feedback on that change. Not a deal breaker, but just feedback we're going to have to consider. 

> Also, I'm curious why the generated class contains a class instance, and doesn't just create the delegate from a static function. Is there a performance difference with using an instance method vs a static function

This code generation pattern was done based on .NET Framework measurements. There it was significantly faster to invoke a delegate pointing to an instance method vs. a static one. The measurements were fairly clear on this change. It's been a while since we've measured to see if this still holds true for .NET Core though.

**Reply by MichalPetryka** [reply]
> Also, I'm curious why the generated class contains a class instance, and doesn't just create the delegate from a static function. Is there a performance difference with using an instance method vs a static function? And would that make a difference to the JIT?

The runtime calls all delegates as instance ones thus it has to use a stub to call static ones.

**Reply by jkotas** [reply]
> Is there any concern with emitting a new static class per each cached delegate?

Yes, each type costs 100's of bytes of memory and 10's microseconds of startup time (these are rough numbers for CoreCLR, for a dummy type).

The current scheme used by Roslyn strikes some tradeoff between static memory footprint and steady state performance, given what's available. It is certainly possible to move this tradeoff towards better steady state performance and worse static footprint, but it will make people who care more about steady state performance happier, and it will make people who care about startup performance less happy. And vice versa.

We can do better: We can make *both* static footprint and steady state performance better.

**Reply by timcassell** [reply]
Quick benchmark on .Net 7:

|                   Type |                Method |     Mean |     Error |    StdDev | Allocated |
|----------------------- |---------------------- |---------:|----------:|----------:|----------:|
| Benchmark_NoMethodImpl |     DelHolderInstCond | 2.452 ns | 0.0927 ns | 0.2718 ns |         - |
| Benchmark_NoMethodImpl |   DelHolderInstNoCond | 1.231 ns | 0.0046 ns | 0.0038 ns |         - |
| Benchmark_NoMethodImpl |   DelHolderStaticCond | 2.813 ns | 0.0771 ns | 0.0683 ns |         - |
| Benchmark_NoMethodImpl | DelHolderStaticNoCond | 2.154 ns | 0.0159 ns | 0.0133 ns |         - |
|        Benchmark_NoOpt |     DelHolderInstCond | 8.859 ns | 0.0175 ns | 0.0146 ns |         - |
|        Benchmark_NoOpt |   DelHolderInstNoCond | 7.752 ns | 0.0177 ns | 0.0166 ns |         - |
|        Benchmark_NoOpt |   DelHolderStaticCond | 9.171 ns | 0.1959 ns | 0.2616 ns |         - |
|        Benchmark_NoOpt | DelHolderStaticNoCond | 7.867 ns | 0.0588 ns | 0.0491 ns |         - |

<details><summary>Code</summary>

```cs
[BenchmarkCategory(nameof(Benchmark_NoOpt))]
public class Benchmark_NoOpt
{
    [MethodImpl(MethodImplOptions.NoOptimization)]
    private void Invoke(Action action) => action();

    private sealed class delHolderInstCond
    {
        public static readonly delHolderInstCond instance = new delHolderInstCond();

        public static Action del;
        internal void method() { }
    }

    [Benchmark]
    [MethodImpl(MethodImplOptions.NoOptimization)]
    public void DelHolderInstCond()
    {
        Invoke(delHolderInstCond.del ??= new Action(delHolderInstCond.instance.method));
    }

    private sealed class delHolderInstNoCond
    {
        public static readonly delHolderInstNoCond instance = new delHolderInstNoCond();

        public static Action del = new Action(instance.method);
        private void method() { }
    }

    [Benchmark]
    [MethodImpl(MethodImplOptions.NoOptimization)]
    public void DelHolderInstNoCond()
    {
        Invoke(delHolderInstNoCond.del);
    }

    private static class delHolderStaticCond
    {
        public static Action del;
        internal static void method() { }
    }

    [Benchmark]
    [MethodImpl(MethodImplOptions.NoOptimization)]
    public void DelHolderStaticCond()
    {
        Invoke(delHolderStaticCond.del ??= new Action(delHolderStaticCond.method));
    }

    private static class delHolderStaticNoCond
    {
        public static Action del = new Action(method);
        private static void method() { }
    }

    [Benchmark]
    [MethodImpl(MethodImplOptions.NoOptimization)]
    public void DelHolderStaticNoCond()
    {
        Invoke(delHolderStaticNoCond.del);
    }
}

[BenchmarkCategory(nameof(Benchmark_NoMethodImpl))]
public class Benchmark_NoMethodImpl
{
    private void Invoke(Action action) => action();

    private sealed class delHolderInstCond
    {
        public static readonly delHolderInstCond instance = new delHolderInstCond();

        public static Action del;
        internal void method() { }
    }

    [Benchmark]
    public void DelHolderInstCond()
    {
        Invoke(delHolderInstCond.del ??= new Action(delHolderInstCond.instance.method));
    }

    private sealed class delHolderInstNoCond
    {
        public static readonly delHolderInstNoCond instance = new delHolderInstNoCond();

        public static Action del = new Action(instance.method);
        private void method() { }
    }

    [Benchmark]
    public void DelHolderInstNoCond()
    {
        Invoke(delHolderInstNoCond.del);
    }

    private static class delHolderStaticCond
    {
        public static Action del;
        internal static void method() { }
    }

    [Benchmark]
    public void DelHolderStaticCond()
    {
        Invoke(delHolderStaticCond.del ??= new Action(delHolderStaticCond.method));
    }

    private static class delHolderStaticNoCond
    {
        public static Action del = new Action(method);
        private static void method() { }
    }

    [Benchmark]
    public void DelHolderStaticNoCond()
    {
        Invoke(delHolderStaticNoCond.del);
    }
}
```

</details>

Event without JIT changes, it seems there could at least be a win in there by eliminating conditional assignments where it's not necessary, especially since they are generated per scope. The conditional assignments should only be generated if they are themselves accessed from a condition, unless it's the only delegate for the entire scope.

**Reply by jaredpar** [reply]
> We can make both static footprint and steady state performance better.

What approach are you thinking of for this? Seems hard to do this alone in the compiler, probably needs to be paired with a runtime change. Or I missing something here?

**Reply by jkotas** [reply]
Yes, we would need new runtime API for Roslyn to use when available. Roslyn does about as good as it can with runtime APIs available today.

**Reply by MichalPetryka** [reply]
> Yes, we would need new runtime API for Roslyn to use when available. Roslyn does about as good as it can with runtime APIs available today.

How would such API look like? Would it take `new Action(obj, pointer)` directly and the runtime would either hoist it to a static field or inline it?

**Reply by jkotas** [reply]
Something like this
```csharp
public class RuntimeHelpers
{
    [Intrinsic]
    // ldftn instruction must immediately precede call of this method. The target method must
    // match the delegate signature and it must be an instance method on the TScope reference. TScope should
    // have no instance fields. Given function pointer can be only mapped to one delegate type using this method.
    static TDelegate GetLambdaSingleton<TScope, TDelegate>(IntPtr ftn) where TScope: class, TDelegate: delegate
    {
        // This is mock implementation. This method is always going to be expanded as JIT intrinsic.
        lock (s_table)
        {
            MethodInfo mi = MapEntryPointToMethodInfo(typeof(TScope), ftn);

            if (s_table.TryGetValue(mi, out Delegate del))
                return (TDelegate)del;

            Delegate ret = typeof(TScope).IsCollectible ? 
                 InternalCreateDelegate(typeof(TDelegate), new TScope(), ftn) :
                 InternalCreateFrozenDelegate(typeof(TDelegate), new-frozen TScope(), ftn);
            s_table.Add(mi, ret);
            return (TDelegate)ret;
        }
    }

    static readonly ConditionalWeakTable<MethodInfo, Delegate> s_table;
}
```

**Reply by jaredpar** [reply]
> TScope should have no instance fields.

That will cause a bit of a refactoring for us. Today a single closure type can contain methods for capturing and `static` lambdas. To meet this constraint, we'll need to break them a part a bit.

**Reply by jkotas** [reply]
This detail is negotiable. For the runtime frozen objects to work, we only need to guarantee that no object reference is ever stored in `TScope` instance. All object reference fields of `TScope` instance used for static lambdas must be null. Also, `TScope` is really only necessary for passing instantiation information to shared generic code. Otherwise, the delegate can be bound to null `this` - no `TScope` instance needed. The details on where it is necessary to create a dummy `TScope` instance may be handled transparently by the runtime.

**Reply by EgorBo** [reply]
To visualize what Jan meant by non-jit friendly pattern for non-capturing lambdas:

![image](https://user-images.githubusercontent.com/523221/232300300-318531e6-1602-40d7-80dd-10ca26582be1.png)

Here we succesfully inline the lambda using our PGO Guarded devirtualization trick (to just `return 42;`), but since the pattern is unfriendly ([mutable static field](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQhAxAHYCuANpQIbCUwAEAJgJYDOdDhhGOSjbIwDCjAN6FGUwf0atSAF0YAVGOwUAKAJSMAvAD5GASVIA3CAGsYGAKwaNCADSMAnjoOMEjANSutAbh4CaRkBeSUTcytbDVsAHnDnRLlFQwAzclIwd3TMsA0AZixnHAAGAMIAXyA==) JIT is not allowed to fold) we're left with a bunch of never-executed (in tier1) code + GDV check that always passes (highlighted by red blocks)

**Reply by CyrusNajmabadi** [reply]
> but since the pattern is unfriendly ([mutable static field](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQhAxAHYCuANpQIbCUwAEAJgJYDOdDhhGOSjbIwDCjAN6FGUwf0atSAF0YAVGOwUAKAJSMAvAD5GASVIA3CAGsYGAKwaNCADSMAnjoOMEjANSutAbh4CaRkBeSUTcytbDVsAHnDnRLlFQwAzclIwd3TMsA0AZixnHAAGAMIAXyA==) JIT is not allowed to fold)

Would it be possible to introduce an attribute that the compiler could emit on this that the theme could then use use to know it was allowed to fold here?

**Reply by MichalPetryka** [reply]
> Would it be possible to introduce an attribute that the compiler could emit on this that the theme could then use use to know it was allowed to fold here?

While it would work with Tiered JIT, it'd still be really messy to detect and it wouldn't work in AOT scenarios (Native AOT is able to interpret static constructors but wouldn't be able to handle this). And it'd still be a breaking change due to blocking modification of those fields, same as making the fields `static readonly`. So it trades working in less scenarios for not allocating Delegate object more eagerly.

**Reply by timcassell** [reply]
Even if the runtime could optimize these delegates that are invoked synchronously, would it still be able to optimize the case of them called asynchronously? I imagine storing a function pointer and invoking it later in the process would still be more efficient than storing a delegate and invoking it later, regardless of what optimizations happen here. Am I wrong?

Either way, I don't think any optimizations that occur here should stop lambda assignments to function pointers.

**Reply by EgorBo** [reply]
Yeah I see how it might work with JIT (although, we'll have to check a custom attribute for all static field loads?) and then jit can treat such static mutable loads as initialized if they're not null at the time of jit compilation, but I have no idea how it could be done on the NAOT side. I like the idea with the new API call more 🙂

**Reply by EgorBo** [reply]
> I imagine storing a function pointer and invoking it later in the process would still be more efficient than storing a delegate and invoking it later, regardless of what optimizations happen here. Am I wrong?

delegate might be even better for your case if it's devirtualized with PGO (we don't devirtualize function pointers currently) e.g.:

![image](https://user-images.githubusercontent.com/523221/232338833-5bec87aa-c82b-4cd0-b595-5660f85896e3.png)

devirtualized and inlined to just `return 42;` you only pay for a quick "is it the most popular delegate here?" check

**Reply by jkotas** [reply]
> Yeah I see how it might work with JIT (although, we'll have to check a custom attribute for all static field loads?)  and then jit can treat such static mutable loads as initialized if they're not null at the time of jit compilation, but I have no idea how it could be done on the NAOT side. I like the idea with the new API call more 

+1. `[ImmutableWhenNonNull]` custom attribute on fields would not be very pay-for-play, and the optimization would not kick in for AOT.

Also, we want to be able to allocate the delegate on frozen heap so that the JIT can embed the delegate address directly into the code without GC handle indirection (ie the codegen would look like the codegen for string literal loads in .NET 8). It means that we want to be able intercept allocation of the delegate as intrinsic, static field load is too late to intercept the delegate allocation. The proposed method would allow us to do that.

**Reply by CyrusNajmabadi** [reply]
Sounds good to me. Thanks for the in depth explanations!

**Reply by jaredpar** [reply]
The reason why we even have an instance on non-capturing lambdas is due to invocation performance reasons. If we implement this frozen object support and JIT primitives does that still hold? Basically, if the delegate points to a static method due we still have the invocation penalty where JIt tends to prefer instance methods? If we could emit these as true `static` members that lowers the refactoring cost for us a bit.

Not a deal breaker to keep the rough structure we have today but partition closures on whether they do / don't capture but it is a moderately expensive item.

**Reply by jkotas** [reply]
> If we implement this frozen object support and JIT primitives does that still hold?

Yes, it still holds.

The delegate callsite assumes instance method. I do not expect that is is going to ever change. If the actual target is a static method, the conversion from instance to static has to happen somewhere, and this conversion needs to be inlined to avoid performance overhead. Roslyn inlines this conversion today by emitting instance methods for static lambdas. Runtime is not able to inline the conversion today and I think would be pretty difficult to teach the runtime to do that.

**Reply by MichalPetryka** [reply]
Should I open an issue for how lambdas are emitted on [dotnet/roslyn](https://github.com/dotnet/roslyn)?

**Reply by jkotas** [reply]
If we want to introduce a new API for this, I think starting with an API proposal issue in dotnet/runtime would be better. Thank you!

**Reply by MichalPetryka** [reply]
Opened https://github.com/dotnet/runtime/issues/85014.
